lib/spack/spack/cmd/buildcache.py

```diff
@@ -642,7 +643,8 @@ def get_concrete_spec(args):
     if spec_str:
         try:
             spec = find_matching_specs(spec_str)[0]
-            spec.concretize()
+            if not spec.concrete:
+                spec.concretize()
```

Comment:
I don't know the full context, but this seems to be covering up a bug. My reasoning is that concretization should reach a fixed point i.e. for a concrete spec `spec.concretize()` should be a no-op.

Reply:
This method could have been called with a spec made from a `.spec.yaml` fetched from an arbitrary binary mirror, and in that case, isn't it true that re-concretization could change it completely?

However, it looks like #20196 is a better way to prevent this behavior from happening in any possible code path.